File: application/lib/models/order.dart
class OrderItemVariant {
  final Map<String, String> attributes;
  final double price;
  final String sku;

  OrderItemVariant({
    required this.attributes,
    required this.price,
    required this.sku,
  });

  factory OrderItemVariant.fromJson(Map<String, dynamic> json) {
    return OrderItemVariant(
      attributes: Map<String, String>.from(json['attributes'] ?? {}),
      price: (json['price'] ?? 0).toDouble(),
      sku: json['sku'] ?? '',
    );
  }
}

class OrderItemProduct {
  final String id;
  final String name;
  final List<String> images;

  OrderItemProduct({
    required this.id,
    required this.name,
    required this.images,
  });

  factory OrderItemProduct.fromJson(Map<String, dynamic> json) {
    return OrderItemProduct(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      images: List<String>.from(json['images'] ?? []),
    );
  }
}

class OrderItem {
  final OrderItemProduct product;
  final OrderItemVariant variant;
  final int quantity;
  final double priceAtOrder;

  OrderItem({
    required this.product,
    required this.variant,
    required this.quantity,
    required this.priceAtOrder,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      product: OrderItemProduct.fromJson(json['product'] ?? {}),
      variant: OrderItemVariant.fromJson(json['variant'] ?? {}),
      quantity: json['quantity'] ?? 0,
      priceAtOrder: (json['priceAtOrder'] ?? 0).toDouble(),
    );
  }

  double get itemTotal => priceAtOrder * quantity;
}

class ShippingAddress {
  final String fullName;
  final String phone;
  final String addressLine1;
  final String addressLine2;
  final String city;
  final String state;
  final String pincode;
  final String country;

  ShippingAddress({
    required this.fullName,
    required this.phone,
    required this.addressLine1,
    required this.addressLine2,
    required this.city,
    required this.state,
    required this.pincode,
    required this.country,
  });

  factory ShippingAddress.fromJson(Map<String, dynamic> json) {
    return ShippingAddress(
      fullName: json['fullName'] ?? '',
      phone: json['phone'] ?? '',
      addressLine1: json['addressLine1'] ?? '',
      addressLine2: json['addressLine2'] ?? '',
      city: json['city'] ?? '',
      state: json['state'] ?? '',
      pincode: json['pincode'] ?? '',
      country: json['country'] ?? 'India',
    );
  }

  String get fullAddress {
    final parts = <String>[
      addressLine1,
      if (addressLine2.isNotEmpty) addressLine2,
      city,
      state,
      pincode,
      country,
    ];
    return parts.join(', ');
  }
}

class Order {
  final String id;
  final List<OrderItem> items;
  final ShippingAddress shippingAddress;
  final String paymentMethod;
  final String paymentStatus;
  final String paymentId;
  final String orderStatus;
  final double totalAmount;
  final String notes;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Order({
    required this.id,
    required this.items,
    required this.shippingAddress,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.paymentId,
    required this.orderStatus,
    required this.totalAmount,
    required this.notes,
    required this.createdAt,
    this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['_id'] ?? '',
      items: (json['items'] as List?)
          ?.map((e) => OrderItem.fromJson(e))
          .toList() ?? [],
      shippingAddress: ShippingAddress.fromJson(json['shippingAddress'] ?? {}),
      paymentMethod: json['paymentMethod'] ?? '',
      paymentStatus: json['paymentStatus'] ?? 'pending',
      paymentId: json['paymentId'] ?? '',
      orderStatus: json['orderStatus'] ?? 'pending',
      totalAmount: (json['totalAmount'] ?? 0).toDouble(),
      notes: json['notes'] ?? '',
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  // Helper methods
  bool get isPending => orderStatus == 'pending';
  bool get isConfirmed => orderStatus == 'confirmed';
  bool get isProcessing => orderStatus == 'processing';
  bool get isShipped => orderStatus == 'shipped';
  bool get isDelivered => orderStatus == 'delivered';
  bool get isCancelled => orderStatus == 'cancelled';

  bool get isPaymentPending => paymentStatus == 'pending';
  bool get isPaymentCompleted => paymentStatus == 'completed';
  bool get isPaymentFailed => paymentStatus == 'failed';

  String get orderStatusDisplay {
    switch (orderStatus) {
      case 'pending':
        return 'Pending';
      case 'confirmed':
        return 'Confirmed';
      case 'processing':
        return 'Processing';
      case 'shipped':
        return 'Shipped';
      case 'delivered':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      default:
        return orderStatus;
    }
  }

  String get paymentMethodDisplay {
    switch (paymentMethod) {
      case 'razorpay':
        return 'Razorpay';
      case 'stripe':
        return 'Stripe';
      case 'cod':
        return 'Cash on Delivery';
      default:
        return paymentMethod;
    }
  }
}

File: src/models/Order.js
import mongoose from 'mongoose';

const OrderItemSchema = new mongoose.Schema({
    // Null only for a custom (off-catalog) POS line — see isCustom below.
    // A catalog line without a product would break restock and P/L silently,
    // so the two fields are validated against each other.
    product: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Product',
        default: null,
        validate: {
            validator: function (value) {
                return this.isCustom || value != null;
            },
            message: 'Order item needs a product unless it is a custom line'
        }
    },
    // Rung up at the counter for something not in the catalog. Such a line has
    // no product, no variant and no stock behind it: nothing to reserve at sale
    // time and nothing to restock on a return. The name/price are whatever the
    // cashier typed, so these are listed for review in admin afterwards.
    isCustom: {
        type: Boolean,
        default: false
    },
    // Store variant details as they were at time of order
    variant: {
        attributes: {
            type: Map,
            of: String
        },
        price: {
            type: Number,
            required: true
        },
        sku: String
    },
    quantity: {
        type: Number,
        required: true,
        min: 1
    },
    // Price at time of order (in case product price changes later)
    priceAtOrder: {
        type: Number,
        required: true
    },
    // Cost snapshot at sale time so P/L does not rewrite history if cost changes
    costAtOrder: {
        type: Number,
        default: null,
        min: 0
    },
    // Name captured at sale time so a renamed product does not rewrite history
    name: {
        type: String,
        default: ''
    },
    // Discount applied to this line only, before any bill-wide discount
    lineDiscount: {
        type: Number,
        default: 0,
        min: 0
    },
    // GST snapshot for this line
    taxRate: {
        type: Number,
        default: 0
    },
    taxAmount: {
        type: Number,
        default: 0
    },
    hsn: {
        type: String,
        default: ''
    },
    // Units sent back through a return
    returnedQuantity: {
        type: Number,
        default: 0,
        min: 0
    }
}, { _id: false });

const RefundSchema = new mongoose.Schema({
    amount: {
        type: Number,
        required: true,
        min: 0
    },
    reason: {
        type: String,
        trim: true,
        maxlength: 300,
        default: ''
    },
    // Which lines came back, and how many of each
    items: [{
        itemIndex: { type: Number, required: true },
        quantity: { type: Number, required: true, min: 1 },
        _id: false
    }],
    restocked: {
        type: Boolean,
        default: true
    },
    processedBy: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        default: null
    },
    createdAt: {
        type: Date,
        default: Date.now
    }
}, { _id: true });

const OrderSchema = new mongoose.Schema({
    // Absent for POS walk-in sales, which have no account behind them.
    user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        default: null
    },
    source: {
        type: String,
        enum: ['web', 'pos'],
        default: 'web'
    },
    // Counter customer behind a POS sale, matched on phone number
    customer: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Customer',
        default: null
    },
    items: {
        type: [OrderItemSchema],
        required: true,
        validate: {
            validator: function (v) {
                return v && v.length > 0;
            },
            message: 'Order must have at least one item'
        }
    },
    shippingAddress: {
        fullName: {
            type: String,
            required: true,
            trim: true
        },
        phone: {
            type: String,
            required: true,
            trim: true
        },
        addressLine1: {
            type: String,
            required: true,
            trim: true
        },
        addressLine2: {
            type: String,
            trim: true,
            default: ''
        },
        city: {
            type: String,
            required: true,
            trim: true
        },
        state: {
            type: String,
            required: true,
            trim: true
        },
        pincode: {
            type: String,
            required: true,
            trim: true
        },
        country: {
            type: String,
            required: true,
            default: 'India',
            trim: true
        }
    },
    paymentMethod: {
        // cash/card/upi are counter payments taken through the POS
        type: String,
        enum: ['razorpay', 'stripe', 'cod', 'cash', 'card', 'upi'],
        required: true
    },
    paymentStatus: {
        type: String,
        enum: ['pending', 'completed', 'failed', 'refunded', 'partially_refunded'],
        default: 'pending'
    },
    paymentId: {
        type: String,
        default: ''
    },
    orderStatus: {
        type: String,
        enum: ['pending', 'confirmed', 'processing', 'shipped', 'delivered', 'cancelled', 'returned', 'partially_returned'],
        default: 'pending'
    },
    totalAmount: {
        type: Number,
        required: true,
        min: 0
    },
    // Coupon tracking
    subtotal: {
        type: Number,
        default: 0
    },
    coupon: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Coupon',
        default: null
    },
    couponCode: {
        type: String,
        default: ''
    },
    discountAmount: {
        type: Number,
        default: 0
    },
    // Sequential invoice serial, allocated once at sale time
    billNumber: {
        type: String,
        default: ''
    },
    // GST totals. Counter sales are intra-state, so tax splits evenly into
    // CGST and SGST; taxBreakup keeps the per-slab rows a tax invoice needs.
    taxAmount: {
        type: Number,
        default: 0
    },
    taxBreakup: [{
        rate: { type: Number, required: true },
        taxable: { type: Number, required: true },
        tax: { type: Number, required: true },
        _id: false
    }],
    // Sum of per-line discounts, kept apart from the bill-wide discountAmount
    lineDiscountTotal: {
        type: Number,
        default: 0
    },
    refunds: {
        type: [RefundSchema],
        default: []
    },
    refundedAmount: {
        type: Number,
        default: 0
    },
    notes: {
        type: String,
        default: '',
        maxlength: 500
    },
    shippingFee: {
        type: Number,
        default: 0
    },
    shipping: {
        shiprocketOrderId: {
            type: String,
            default: ''
        },
        shipmentId: {
            type: String,
            default: ''
        },
        awbCode: {
            type: String,
            default: ''
        },
        courierName: {
            type: String,
            default: ''
        },
        courierId: {
            type: Number,
            default: null
        },
        label: {
            type: String,
            default: ''
        },
        trackingUrl: {
            type: String,
            default: ''
        },
        estimatedDelivery: {
            type: Date,
            default: null
        },
        status: {
            type: String,
            default: ''
        },
        lastUpdate: {
            type: Date,
            default: null
        }
    }
}, {
    timestamps: true
});

// Indexes for faster queries
OrderSchema.index({ user: 1, createdAt: -1 });
OrderSchema.index({ source: 1, createdAt: -1 });
OrderSchema.index({ orderStatus: 1 });
OrderSchema.index({ paymentStatus: 1 });

export default mongoose.models.Order || mongoose.model('Order', OrderSchema);
